File: day3/haskell/part2.hs
import System.Environment (getArgs)
import Data.Char (digitToInt)
import Data.List (sortBy)
import Data.Ord (comparing)

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "input.txt" else head args
    content <- readFile filename
    let total = sum (map maxJoltage12 (lines content))
    print total

maxJoltage12 :: String -> Integer
maxJoltage12 line = read result
  where
    n = length line
    k = 12
    result = select k (-1) ""

    select 0 _ acc = acc
    select remaining lastPos acc =
        let searchEnd = n - k + (k - remaining)
            (bestDigit, bestPos) = maximum [(line !! j, j) | j <- [lastPos + 1 .. searchEnd]]
        in select (remaining - 1) bestPos (acc ++ [bestDigit])

File: day3/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func maxJoltage12(line string) int64 {
	n := len(line)
	k := 12 // digits to pick

	result := ""
	lastPos := -1

	for i := 0; i < k; i++ {
		// For digit i, need to leave k-1-i digits after
		// So search from lastPos+1 to n-k+i
		bestDigit := byte('0')
		bestPos := -1

		for j := lastPos + 1; j <= n-k+i; j++ {
			if line[j] > bestDigit {
				bestDigit = line[j]
				bestPos = j
			}
		}

		result = result + string(bestDigit)
		lastPos = bestPos
	}

	num, _ := strconv.ParseInt(result, 10, 64)
	return num
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, _ := os.Open(filename)
	defer file.Close()

	var sum int64 = 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		sum = sum + maxJoltage12(line)
	}

	fmt.Println(sum)
}
